### PR TITLE
Scala tutorial using JDBC

### PR DESCRIPTION
**Pull request by ennru**
Switches to use of PostgreSQL with Akka Persistence JDBC and ScalikeJDBC for database access in Projections.

### CODE REVIEW

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ItemPopularityRepository.scala [hunk 1]
```diff
@@ -2,45 +2,43 @@
...
+  override def update(itemId: String, delta: Int): Unit = {
+    // TODO hook into transaction
+    DB.localTx { implicit session =>
+      // This uses the PostgreSQL `ON CONFLICT` feature
+      // Alternatively, this can be implemented by first issuing the `UPDATE`
+      // and checking for the updated rows count. If no rows got updated issue
+      // the `INSERT` instead.
+      sql"""
+           INSERT INTO item_popularity (itemid, count) VALUES ($itemId, $delta)
+           ON CONFLICT (itemid) DO UPDATE SET count = item_popularity.count + $delta
```
Comment (patriknw) [inline docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ItemPopularityRepository.scala:30]: does the `sql` interpolation take care of using the variables in prepared statements?

**Review by patriknw — COMMENTED** [review]
looking very good

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ItemPopularityRepository.scala (hunk 1 above)
Comment (ennru) [reply docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ItemPopularityRepository.scala:30]: Yes, that part works very nicely.

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/repository/ScalikeJdbcSession.scala [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+import akka.projection.jdbc.JdbcSession
+import scalikejdbc._
+
+import java.sql.Connection
+
+final class ScalikeJdbcSession extends JdbcSession {
+  private implicit val db: DB = DB.connect()
+
+  override def withConnection[Result](
+      func: Function[Connection, Result]): Result = {
```
Comment (octonato) [inline docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/repository/ScalikeJdbcSession.scala:13]: hmm, we did the JDBC api with Java in mind, but now we are using it from Scala. Weird.

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/test/scala/shopping/cart/CreateTableTestUtils.scala [hunk 3]
```diff
@@ -1,39 +1,64 @@
...
+          _ <- SchemaUtils.createIfNotExists()
+          _ <- JdbcProjection.dropOffsetTableIfExists(() => scalikeJdbcSession)
+          _ <- JdbcProjection.createOffsetTableIfNotExists(() =>
+            scalikeJdbcSession)
+          _ <- SchemaUtils.applyScript(
+            fromFileAsString("ddl-scripts/drop_user_tables.sql"))
+          _ <- SchemaUtils.applyScript(
+            fromFileAsString("ddl-scripts/create_user_tables.sql"))
+        } yield Done,
+        30.seconds)
```
Comment (octonato) [inline docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/test/scala/shopping/cart/CreateTableTestUtils.scala:48]: 👍 
We need to port this to the Java variant. I will do it.

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/repository/ScalikeJdbcSession.scala [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    DB.autoCommitWithConnection(func(_))
```
Comment (octonato) [inline docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/repository/ScalikeJdbcSession.scala:14]: Without the means to get a connection, commit and close it ourselves we won't be able to have exactly once. 

We should better have the answer as soon as possible before we continue investing in ScalaLikeJDBC

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCartServiceImpl.scala [hunk 5]
```diff
@@ -107,7 +107,8 @@ class ShoppingCartServiceImpl(
   // tag::getItemPopularity[]
   override def getItemPopularity(in: proto.GetItemPopularityRequest)
       : Future[proto.GetItemPopularityResponse] = {
-    itemPopularityRepository.getItem(in.itemId).map {
+    // TODO proper async
+    Future(itemPopularityRepository.getItem(in.itemId)).map {
```
Comment (octonato) [inline docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCartServiceImpl.scala:111]: In the Java variant, we are using Akka's blocking dispatcher. But I realise that this is wrong. 

The repository will pick a connection from the pool and we should pick a thread from `akka.projection.jdbc.blocking-jdbc-dispatcher.thread-pool-executor.fixed-pool-size`. 

If connections are used by threads outside the `blocking-jdbc-dispatcher`, then we can have a projection thread waiting for the connection pool to release a connection. 

So basically, whenever we use the repo in async fashion, we need to schedule it on the `blocking-jdbc-dispatcher`. 

That's a little error prone and the user will have to be vigilant. I wish we had something to enforce this.

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ItemPopularityRepository.scala [hunk 6]
```diff
@@ -2,45 +2,43 @@
...
-  override def update(itemId: String, delta: Int): Future[Done] = {
-    session.executeWrite(
-      s"UPDATE $keyspace.$popularityTable SET count = count + ? WHERE item_id = ?",
-      java.lang.Long.valueOf(delta),
-      itemId)
+class ItemPopularityRepositoryImpl() extends ItemPopularityRepository {
+
+  override def update(itemId: String, delta: Int): Unit = {
+    // TODO hook into transaction
+    DB.localTx { implicit session =>
```
Comment (octonato) [inline docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ItemPopularityRepository.scala:23]: This is looking closer to the Spring impl. Find a connection somewhere in a magic static place (or thread local).

**Review by octonato — COMMENTED** [review]
@ennru, I left some comments. 

I'm a little concerned with the level of control we will have of connection and threads and the amount of burden we will put on users.

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ItemPopularityRepository.scala [hunk 7]
```diff
@@ -3,24 +3,27 @@
...
 class ItemPopularityRepositoryImpl() extends ItemPopularityRepository {
 
-  override def update(itemId: String, delta: Int): Unit = {
-    // TODO hook into transaction
-    DB.localTx { implicit session =>
+  override def update(
+      session: ScalikeJdbcSession,
+      itemId: String,
+      delta: Int): Unit = {
+    session.db.withinTx { implicit session =>
```
Comment (patriknw) [inline docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ItemPopularityRepository.scala:26]: nice

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/repository/ScalikeJdbcSession.scala [hunk 8]
```diff
@@ -2,21 +2,39 @@ package shopping.cart.repository
...
+  override def commit(): Unit = {
+    db.commit()
+    log.debug("committed {}", db)
+  }
 
-  override def rollback(): Unit = ()
+  override def rollback(): Unit = db.rollback()
 
-  override def close(): Unit = db.close()
+  override def close(): Unit = () //db.close() //db.close()
```
Comment (patriknw) [inline docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/repository/ScalikeJdbcSession.scala:39]: ?

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/test/resources/persistence-test.conf [hunk 9]
```diff
@@ -1,5 +1,16 @@
 jdbc-connection-settings {
-  url = "jdbc:postgresql://localhost:5433/shopping-cart-test?reWriteBatchedInserts=true"
+  url = "jdbc:postgresql://localhost:5433/shopping-cart-test"
```
Comment (patriknw) [inline docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/test/resources/persistence-test.conf:2]: do we know if that `?reWriteBatchedInserts=true` should be included or not? I think it comes from akka-persistence-jdbc documentation

**Review by patriknw — COMMENTED** [review]
tx stuff is looking good

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/test/scala/shopping/cart/CreateTableTestUtils.scala [hunk 10]
```diff
@@ -1,39 +1,66 @@
...
+    val scalikeJdbcSession = new ScalikeJdbcSession()
+    try {
+      // ok to block here, main thread
+      Await.result(
+        for {
+          _ <- SchemaUtils.dropIfExists()
+          _ <- SchemaUtils.createIfNotExists()
+          _ <- JdbcProjection.dropOffsetTableIfExists(() => scalikeJdbcSession)
+          _ <- JdbcProjection.createOffsetTableIfNotExists(() =>
+            scalikeJdbcSession)
```
Comment (patriknw) [inline docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/test/scala/shopping/cart/CreateTableTestUtils.scala:44]: This is wrong usage ScalikeJdbcSession and that seems to cause problem with the connections. A ScalikeJdbcSession instance shouldn't be used from multiple threads, and the utilities dropOffsetTableIfExists and reateOffsetTableIfNotExists will close it.

It should be a new instance for each of them.

```
    // ok to block here, main thread
    Await.result(
      for {
        _ <- SchemaUtils.dropIfExists()
        _ <- SchemaUtils.createIfNotExists()
        _ <- JdbcProjection.dropOffsetTableIfExists(() => new ScalikeJdbcSession())
        _ <- JdbcProjection.createOffsetTableIfNotExists(() => new ScalikeJdbcSession())
        _ <- SchemaUtils.applyScript(
          fromFileAsString("ddl-scripts/drop_user_tables.sql"))
        _ <- SchemaUtils.applyScript(
          fromFileAsString("ddl-scripts/create_user_tables.sql"))
      } yield Done,
      30.seconds)
```

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/repository/ScalikeJdbcSession.scala [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    } finally {
+      session.close()
+    }
+  }
+}
+
+final class ScalikeJdbcSession extends JdbcSession {
+  private val log = LoggerFactory.getLogger(getClass)
+
+  val db: DB = DB.connect()
```
Comment (patriknw) [inline docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/repository/ScalikeJdbcSession.scala:24]: I think we should use `d.autoClose(false)` here, because otherwise it can only be used for one operation, see "Reusing same DB instance several times" in http://scalikejdbc.org/documentation/connection-pool.html

**Review by patriknw — COMMENTED** [review]
fix for connection problem

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/repository/ScalikeJdbcSession.scala [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+  override def commit(): Unit = {
+    db.commit()
+    log.debug("committed {}", db)
+  }
+
+  override def rollback(): Unit = db.rollback()
+
+  override def close(): Unit = {
+    if (db.toString.contains("org.apache.commons.dbcp2.PoolingDataSource")) {
```
Comment (patriknw) [inline docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/repository/ScalikeJdbcSession.scala:42]: and this should be removed again

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/test/scala/shopping/cart/CreateTableTestUtils.scala (hunk 10 above)
Comment (ennru) [reply docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/test/scala/shopping/cart/CreateTableTestUtils.scala:44]: Thanks. I was digging at the wrong end.

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/resources/persistence.conf [hunk 13]
```diff
@@ -1,45 +1,93 @@
...
+  # The `poolMaxSize` should not be smaller than the `blocking-jdbc-dispatcher` pool
+  # configured in `akka.projection`.
+  poolMaxSize = 15
+  # Fail fast if the pool can't instantly provide a connection.
+  poolConnectionTimeoutMillis = 20
+  poolValidationQuery = "select 1 as one"
+  # ScalikeJDBC comes with this connection pool implementation
+  # As Akka Persistence JDBC with Slick uses Hikari it makes sense to implement
+  # a Hikari pool for ScalikeJDBC
+  poolFactoryName = "commons-dbcp2"
```
Comment (octonato) [inline docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/resources/persistence.conf:87]: It will be good if we align it with the java variant. There we have now this:

```hocon

# this settings is used to configure the Spring Data layer and the akka-persistence-jdbc plugin
jdbc-connection-settings {
  # the following properties are used to configure the
  # Hikari connection pool used on the read-side (akka-projections)
  connection-pool {
    # How many connections should be available to from the pool?
    # it's recommended to use the same value used by the blocking-jdbc-dispatcher (see above)
    max-pool-size = ${akka.projection.jdbc.blocking-jdbc-dispatcher.thread-pool-executor.fixed-pool-size}

    # How long should we wait (in millis) before it times out?
    # In a normal scenario, we should always be able to get a connection
    # If we got a thread from the blocking-jdbc-dispatcher, we should be able to get a connection.
    # If for some reason the pool can't provide a connection, it's better to let it crash and liberate the current thread.
    # Hence the low timout (note, 250 is lowest value hikari accepts)
    timeout = 250ms
  }

}
```

We then use the `jdbc-connection-settings` block to configure a Hikari datasource and we let Spring pick it as the data source to use. It would be nice to investigate if we can do the same with ScalikeJdbc. 

So, instead of loading from config, we could create a DataSource and tell ScalikeJdbc to use it. That will give us better control on how we load the properties.

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCartServiceImpl.scala [hunk 14]
```diff
@@ -107,7 +106,15 @@ class ShoppingCartServiceImpl(
   // tag::getItemPopularity[]
   override def getItemPopularity(in: proto.GetItemPopularityRequest)
       : Future[proto.GetItemPopularityResponse] = {
-    itemPopularityRepository.getItem(in.itemId).map {
+    // TODO proper async
+    Future {
```
Comment (octonato) [inline docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCartServiceImpl.scala:110]: TODO leftover?

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/dbaccess/ItemPopularityRepository.scala [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+      itemId: String): Option[Long] = {
+    if (session.db.isTxAlreadyStarted) {
+      session.db.withinTx { implicit session =>
+        select(itemId)
+      }
+    } else {
+      session.db.readOnly { implicit session =>
+        select(itemId)
+      }
+    }
```
Comment (octonato) [inline docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/dbaccess/ItemPopularityRepository.scala:41]: what happens if we remove the if/else and always run it using `session.db.withinTx`? Don't they have an API to deal with re-entrant transactions?

**Review by octonato — APPROVED** [review]
@ennru, I left some more comments and thoughts. Otherwise I think it's good to go.

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/resources/persistence.conf (hunk 13 above)
Comment (ennru) [reply docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/resources/persistence.conf:87]: Let's do that in separate PR.

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCartServiceImpl.scala (hunk 14 above)
Comment (ennru) [reply docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCartServiceImpl.scala]: Fixed in [`02bdb18` (#459)](https://github.com/akka/akka-platform-guide/pull/459/commits/02bdb1887d7cb5d2720bea57405511c183937af3)

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/dbaccess/ItemPopularityRepository.scala (hunk 15 above)
Comment (ennru) [reply docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/dbaccess/ItemPopularityRepository.scala]: `withinTx` expects a transaction being active and fails otherwise.

File: docs-source/docs/modules/how-to/pages/health-checks.adoc [hunk 16]
```diff
@@ -53,7 +53,13 @@ Akka Persistence Cassandra includes a check to validate connectivity, as the ser
 .src/main/resources/application.conf
 [source,hocon]
 ----
-include::microservices-tutorial:example$shopping-cart-service-scala/src/main/resources/persistence.conf[tag=healthchecks-for-cassandra]
```
Comment (patriknw) [inline docs-source/docs/modules/how-to/pages/health-checks.adoc:56]: this exists again in 
```
include::how-to:example$shopping-cart-service-cassandra-scala/src/main/resources/persistence.conf[tag=healthchecks-for-cassandra]
```

File: scripts/copy-identical-files.sh [hunk 17]
```diff
@@ -404,6 +411,23 @@ cp ${SRC} ${tutorial_root}/02-shopping-cart-service-java/src/main/java/shopping/
 cp ${SRC} ${tutorial_root}/01-shopping-cart-service-java/src/main/java/shopping/cart/repository/
 cp ${SRC} ${tutorial_root}/shopping-cart-service-java/src/main/java/shopping/cart/repository/
 
+
+declare SRC="${tutorial_root}/00-shopping-cart-service-scala/src/main/scala/shopping/cart/repository/DBsFromConfig.scala"
+cp ${SRC} ${tutorial_root}/05-shopping-cart-service-scala/src/main/scala/shopping/cart/repository/
+cp ${SRC} ${tutorial_root}/04-shopping-cart-service-scala/src/main/scala/shopping/cart/repository/
+cp ${SRC} ${tutorial_root}/03-shopping-cart-service-scala/src/main/scala/shopping/cart/repository/
+cp ${SRC} ${tutorial_root}/02-shopping-cart-service-scala/src/main/scala/shopping/cart/repository/
+cp ${SRC} ${tutorial_root}/01-shopping-cart-service-scala/src/main/scala/shopping/cart/repository/
```
Comment (patriknw) [inline scripts/copy-identical-files.sh:420]: not important, but different order than other things that are copied from 00

File: docs-source/docs/modules/microservices-tutorial/pages/projection-query.adoc [hunk 18]
```diff
@@ -122,10 +128,10 @@ Scala::
 .src/main/scala/shopping/cart/ItemPopularityProjectionHandler.scala:
 [source,scala,indent=0]
 ----
-include::example$04-shopping-cart-service-scala/src/main/scala/shopping/cart/ItemPopularityProjectionHandler.scala[tags=handler]
+include::example$shopping-cart-service-scala/src/main/scala/shopping/cart/ItemPopularityProjectionHandler.scala[tags=handler]
```
Comment (patriknw) [inline docs-source/docs/modules/microservices-tutorial/pages/projection-query.adoc:131]: the intention was that the includes should correspond to the corresponding source, which is 04 for this part

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/repository/DBsFromConfig.scala [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+
+  /**
+   * Initiate the ScalikeJDBC connection pool configuration and shutdown.
+   */
+  def init(system: ActorSystem[_]): Unit = {
+    val scalikeJdbc: DBsFromConfig = fromConfig(system.settings.config)
+    system.whenTerminated.map { _ =>
+      scalikeJdbc.closeAll()
+    }(scala.concurrent.ExecutionContext.Implicits.global)
+  }
```
Comment (patriknw) [inline docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/repository/DBsFromConfig.scala:31]: good

File: docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/Main.scala [hunk 20]
```diff
@@ -33,23 +33,15 @@ object Main {
 class Main(context: ActorContext[Nothing])
     extends AbstractBehavior[Nothing](context) {
   val system = context.system
+
+  DBsFromConfig.init(system)
```
Comment (patriknw) [inline docs-source/docs/modules/microservices-tutorial/examples/shopping-cart-service-scala/src/main/scala/shopping/cart/Main.scala:37]: 👍

**Review by patriknw — APPROVED** [review]
LGTM, excellent

**Comment by patriknw** [comment]
🎉
